File: src/TypeFormatter/OptionalTypeFormatter.ts
import type { Definition } from "../Schema/Definition.js";
import type { SubTypeFormatter } from "../SubTypeFormatter.js";
import type { BaseType } from "../Type/BaseType.js";
import { OptionalType } from "../Type/OptionalType.js";
import type { GetDefinitionOptions } from "../TypeFormatter.js";
import type { TypeFormatter } from "../TypeFormatter.js";

export class OptionalTypeFormatter implements SubTypeFormatter {
    public constructor(protected childTypeFormatter: TypeFormatter) {}

    public supportsType(type: BaseType): boolean {
        return type instanceof OptionalType;
    }
    public getDefinition(type: OptionalType, options?: GetDefinitionOptions): Definition {
        return this.childTypeFormatter.getDefinition(type.getType(), options);
    }
    public getChildren(type: OptionalType): BaseType[] {
        return this.childTypeFormatter.getChildren(type.getType());
    }
}

File: internal/formatter/optional_type_formatter.go
package formatter

import (
	"github.com/vega/ts-json-schema-generator/internal/schema"
	"github.com/vega/ts-json-schema-generator/internal/types"
)

// OptionalTypeFormatter mirrors src/TypeFormatter/OptionalTypeFormatter.ts.
type OptionalTypeFormatter struct {
	childTypeFormatter TypeFormatter
}

func NewOptionalTypeFormatter(childTypeFormatter TypeFormatter) *OptionalTypeFormatter {
	return &OptionalTypeFormatter{childTypeFormatter: childTypeFormatter}
}

func (f *OptionalTypeFormatter) SupportsType(t types.Type) bool {
	_, ok := t.(*types.OptionalType)
	return ok
}

func (f *OptionalTypeFormatter) GetDefinition(t types.Type) *schema.Definition {
	return f.childTypeFormatter.GetDefinition(t.(*types.OptionalType).Type)
}

func (f *OptionalTypeFormatter) GetChildren(t types.Type) []types.Type {
	return f.childTypeFormatter.GetChildren(t.(*types.OptionalType).Type)
}
